src/llvm-late-gc-lowering.cpp

```diff
@@ -0,0 +1,1351 @@
+                    }
+                    assert(it != Def->getParent()->end());
+                }
+            }
+        }
+    }
+}
+
+/* For chordal interference graphs, this class gives the verticies in a (reverse
+ * - depending on definition) perfect elimination ordering, in such a way that
+ * greedy coloring gives an optimial coloring. Since our roots are in SSA form,
+ * the interference should be chordal.
+ */
+struct PEOIterator {
+    struct Element {
+        unsigned weight;
+        unsigned pos;
+    };
+    std::vector<Element> Elements;
+    std::vector<std::vector<int>> Levels;
+    const std::vector<std::vector<int>> &Neighbors;
+    PEOIterator(const std::vector<std::vector<int>> &Neighbors) : Neighbors(Neighbors) {
+        // Initialize State
+        std::vector<int> FirstLevel;
+        for (unsigned i = 0; i < Neighbors.size(); ++i) {
+            FirstLevel.push_back(i);
+            Element E{0, i};
+            Elements.push_back(E);
+        }
+        Levels.push_back(FirstLevel);
+    }
+    int next() {
+        // Find the element in the highest bucket
+        int NextElement = -1;
+        while (NextElement == -1 && !Levels.empty()) {
+            std::vector<int> &LastLevel = Levels.back();
+            while (NextElement == -1 && !LastLevel.empty()) {
+                NextElement = LastLevel.back();
+                LastLevel.pop_back();
+            }
+            if (LastLevel.empty())
+                Levels.pop_back();
+        }
+        if (NextElement == -1)
+            return NextElement;
+        // Make sure not to try to re-use this later.
+        Elements[NextElement].weight = (unsigned)-1;
+        // Raise neighbors
+        for (int Neighbor : Neighbors[NextElement]) {
+            Element &NElement = Elements[Neighbor];
+            // Already processed. Don't re-enqueue
+            if (NElement.weight == (unsigned)-1)
+                continue;
+            // Kill old queue position
+            Levels[NElement.weight][NElement.pos] = -1;
+            // Raise the neighbor to the next level.
+            NElement.weight += 1;
+            if (NElement.weight >= Levels.size())
+                Levels.push_back(std::vector<int>{});
+            Levels[NElement.weight].push_back(Neighbor);
+            NElement.pos = Levels[NElement.weight].size()-1;
+        }
+        // As an enhancement, we might want to periodically compactify the whole
+        // data structure. This could be done here.
+        return NextElement;
+    }
+};
+
+std::vector<int> LateLowerGCFrame::ColorRoots(const State &S) {
+    std::vector<int> Colors;
+    for (int i = 0; i <= S.MaxPtrNumber; ++i)
+        Colors.push_back(-1);
+    PEOIterator Ordering(S.Neighbors);
+    /* Greedy coloring */
+    int ActiveElement = 1;
+    int MaxAssignedColor = -1;
+    while ((ActiveElement = Ordering.next()) != -1) {
+        assert(Colors[ActiveElement] == -1);
+        BitVector UsedColors(MaxAssignedColor + 2, false);
```

Comment:
I think it might be good to move this malloc outside the while loop and just reset its size here (`UsedColors.reset().resize(MaxAssignedColor + 2)`)